src/com/phonebookservice/util/MyArrayList.java

```diff
@@ -0,0 +1,407 @@
+    }
+
+    /**
+     * Removes the first occurrence of the specified element from this list, if
+     * it is present (optional operation).
+     */
+    @Override
+    public boolean remove(final Object element) {
+
+        if (element == null) {
+            for (int i = 0; i < this.currentIndex; i++) {
+                if (this.data[i] == null) {
+                    this.shiftRemove(i);
+                    return true;
+                }
+            }
+        } else {
+            for (int i = 0; i < currentIndex; i++) {
+                if (element.equals(this.data[i])) {
+                    this.shiftRemove(i);
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Removes the element at the specified index in this list.
+     *
+     * @param index the index.
+     */
+    @Override
+    public T remove(final int index) {
+        this.checkIndex(index);
+        T object = this.data[index];
+        this.remove(object);
+        return object;
+    }
+
+    /**
+     * Removes from this list all of its elements that are contained in the
+     * specified collection (optional operation).
+     */
+    @Override
+    public boolean removeAll(final Collection<?> collection) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Retains only the elements in this list that are contained in the
+     * specified collection (optional operation).
+     */
+    @Override
+    public boolean retainAll(final Collection<?> collection) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Checks if this list contains all of the elements of the collection.
+     *
+     * @param collection the collection.
+     * @return true if this list contains all of the elements of the collection.
+     */
+    @Override
+    public boolean containsAll(final Collection<?> collection) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * increase the size of data[].
+     */
+    private void increaseSize() {
+        int newSize = this.data.length * 2;
+        this.data = Arrays.copyOf(this.data, newSize);
+    }
+
+    private void checkIndexToAdd(final int i) {
+        if (i > this.currentIndex || i < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void checkIndex(final int i) {
+        if (i >= this.currentIndex || i < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private T[] copyArray(final T[] data, final int index) {
+        final T[] array = (T[]) new Object[index];
+
+        for (int i = 0; i < index; i++) {
+            array[i] = data[i];
+        }
+        return array;
+    }
+
+    private void shiftRemove(final int i) {
+        for (int j = i; j < this.currentIndex; j++) {
+            this.data[i] = this.data[i + 1];
+        }
+        this.currentIndex--;
+        this.data[this.currentIndex] = null;
```

Comment:
why don't do 
```
this.data[--this.currentIndex] = null;
```

Question :- 
What is the difference between this.data[--this.currentIndex] = null; and this.data[this.currentIndex--] = null;?

Reply:
this.data[--this.currentIndex] = null means to decrement the index, then set data with this index to null.
this.data[this.currentIndex--] = null means setting data with current index to null, then decrement the index.